### PR TITLE
feat(net): get external IPv4 from libp2p

### PR DESCRIPTION
**Pull request by 317787106**
**What does this PR do?**

- get external IPv4 from llibp2p instead of external url

**Why are these changes required?**

**This PR has been tested by:**
- Unit Tests
- Manual Testing

**Follow up**

**Extra details**

### CODE REVIEW

File: framework/src/test/java/org/tron/common/config/args/ArgsTest.java [hunk 1]
```diff
@@ -45,4 +47,18 @@ public void testConfig() {
     Assert.assertEquals(rateLimiter.getHttpMap().size(), 1);
     Assert.assertEquals(rateLimiter.getRpcMap().size(), 0);
   }
+
+  @Test
+  public void testHelpMessage() {
```
Comment (halibobo1205) [inline framework/src/test/java/org/tron/common/config/args/ArgsTest.java:52]: Please submit the tests to cover  this PR, if it is simply to increase coverage, it is recommended to submit a new PR.

File: framework/src/main/java/org/tron/core/config/args/Args.java [hunk 2]
```diff
@@ -1545,40 +1544,14 @@ private static void bindIp(final com.typesafe.config.Config config) {
     }
   }
 
-  private static void externalIp(final com.typesafe.config.Config config) {
+  private static void externalIp(final com.typesafe.config.Config config, P2pConfig p2pConfig) {
```
Comment (chengtx01) [inline framework/src/main/java/org/tron/core/config/args/Args.java:1547]: I think the second parameter is redundant. Can it be removed?

File: framework/src/test/java/org/tron/common/config/args/ArgsTest.java (hunk 1 above)
Comment (317787106) [reply framework/src/test/java/org/tron/common/config/args/ArgsTest.java:52]: i will split this pr later.

File: common/src/main/java/org/tron/common/parameter/CommonParameter.java [hunk 3]
```diff
@@ -177,21 +180,21 @@ public class CommonParameter {
...
-  public long nodeP2pPingInterval;
+//  @Getter
+//  @Setter
+//  public boolean nodeDiscoveryPublicHomeNode;
+//  @Getter
+//  @Setter
+//  public long nodeDiscoveryPingTimeout;
+//  @Getter
+//  @Setter
+//  public long nodeP2pPingInterval;
```
[the remark this answers is not in the document]
Comment (317787106) [reply common/src/main/java/org/tron/common/parameter/CommonParameter.java:191]: I will split this pr later.

File: framework/src/test/java/org/tron/core/ShieldWalletTest.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    PrivateShieldedTRC20Parameters.Builder builder = PrivateShieldedTRC20Parameters.newBuilder();
+    try {
+      JsonFormat.merge(parameter, builder, false);
+    } catch (ParseException e) {
+      Assert.fail();
+    }
+
+    try {
+      ShieldedTRC20Parameters shieldedTRC20Parameters = wallet1.createShieldedContractParameters(
+          builder.build());
```
[the remark this answers is not in the document]
Comment (halibobo1205) [reply framework/src/test/java/org/tron/core/ShieldWalletTest.java:241]: Mockito.when(wallet1.getShieldedContractScalingFactor(...), but actual is wallet1.createShieldedContractParameters(....)?

File: framework/src/main/java/org/tron/core/config/args/Args.java (hunk 2 above)
Comment (317787106) [reply framework/src/main/java/org/tron/core/config/args/Args.java]: @chengtx01  Yes, i have already removed it.

File: framework/src/test/java/org/tron/core/ShieldWalletTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (317787106) [reply framework/src/test/java/org/tron/core/ShieldWalletTest.java:241]: No, i debug this test, it jumps into function createShieldedContractParameters and then return with do nothing.

File: framework/src/test/java/org/tron/core/WalletTest.java [hunk 5]
```diff
@@ -296,9 +297,17 @@ private static Block getBuildBlock(long timestamp, long num, long witnessId,
 
   private void buildAssetIssue() {
     AssetIssueContract.Builder builder = AssetIssueContract.newBuilder();
+    builder.setOwnerAddress(ByteString.copyFromUtf8("Address1"));
```
Comment (halibobo1205) [inline framework/src/test/java/org/tron/core/WalletTest.java:300]: Please don't add non-relevant tests.

File: common/src/main/java/org/tron/core/Constant.java [hunk 6]
```diff
@@ -284,7 +284,7 @@ public class Constant {
   public static final String NODE_DISCOVERY_BIND_IP = "node.discovery.bind.ip";
 
   public static final String NODE_DISCOVERY_EXTERNAL_IP = "node.discovery.external.ip";
-  public static final String AMAZONAWS_URL = "http://checkip.amazonaws.com";
+  //public static final String AMAZONAWS_URL = "http://checkip.amazonaws.com";
```
Comment (halibobo1205) [inline common/src/main/java/org/tron/core/Constant.java:287]: I'd suggest just deleting the useless code.

File: framework/src/test/java/org/tron/core/WalletTest.java (hunk 5 above)
Comment (halibobo1205) [reply framework/src/test/java/org/tron/core/WalletTest.java:300]: Please follow the PR submission guidelines.

**Review by chengtx01 — APPROVED** [review]

File: framework/src/test/java/org/tron/core/config/args/ArgsTest.java [hunk 7]
```diff
@@ -112,5 +123,29 @@ public void get() {
...
+    Method method1 = Args.class.getDeclaredMethod("bindIp", Config.class);
+    method1.setAccessible(true);
+    method1.invoke(Args.class, config3);
+
+    Method method2 = Args.class.getDeclaredMethod("externalIp", Config.class);
+    method2.setAccessible(true);
+    method2.invoke(Args.class, config3);
+
+    Assert.assertNotEquals("127.0.0.1", CommonParameter.getInstance().getNodeDiscoveryBindIp());
+    Assert.assertNotEquals("46.168.1.1", CommonParameter.getInstance().getNodeExternalIp());
```
Comment (halibobo1205) [inline framework/src/test/java/org/tron/core/config/args/ArgsTest.java:148]: `"46.168.1.1"`: What kind of node is this IP?

Comment (317787106) [reply framework/src/test/java/org/tron/core/config/args/ArgsTest.java:148]: 46.168.1.1 is the configed node.discovery.external.ip in Constant.TEST_CONF.
127.0.0.1 s the configed node.discovery.bind.ip in Constant.TEST_CONF.

Comment (halibobo1205) [reply framework/src/test/java/org/tron/core/config/args/ArgsTest.java:148]: Then it is recommended to get the variables from the Constant.TEST_CONF

Comment (317787106) [reply framework/src/test/java/org/tron/core/config/args/ArgsTest.java]: I have updated this testcase.

**Review by halibobo1205 — APPROVED** [review]

**Pull request merged**